### PR TITLE
rec: more sophisticated cname loop detection.

### LINKED ISSUE #9153: rec: CNAME to self from cache doesn't instantly give up
- Program: Recursor
 - Issue type: Bug report

### Short description
When the Recursor encounters a self-referential CNAME (www.example.com. IN CNAME www.example.com.), when it's fresh, it immediately detects the condition, logs "status=got a CNAME referral to self, returning SERVFAIL" and returns SERVFAIL.

When the CNAME is from the cache, it keeps trying, eventually reaches the max depth, gives up, and returns SERVFAIL.

This doesn't do much harm, except for using some CPU cycles, putting the CNAME in the response 8 times, and polluting your logs if you have tracing on.

### Environment
 - Operating system: Ubuntu 16.04
 - Software version: 4.4.0~alpha1+master.127.gbcbc0a729-1pdns.xenial
 - Software source: PowerDNS repository

### Steps to reproduce
[Edit: The domain was fixed, so this no longer reproduces it.]
1. `dig www.les-play.com`
2. `dig www.les-play.com`

(My version of dig has cookies, so it bypasses the packet cache.)

### Expected behaviour
Immediate SERVFAIL, 0-1 CNAME records in response.

### Actual behaviour
More complicated processing, 8 CNAME records in response.

### Other information
`trace-regex 'les-play\.com\.$'` log: https://mn0.us/nET3XYU21r-_BZ8Z_GnrHfu/syslog.txt (2 MB, 14,000 lines)

```
$ dig www.les-play.com

; <<>> DiG 9.17.1-1+ubuntu16.04.1+isc+3-Ubuntu <<>> www.les-play.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 51869
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.les-play.com.              IN      A

;; ANSWER SECTION:
www.les-play.com.       300     IN      CNAME   www.les-play.com.

;; Query time: 39 msec
;; SERVER: ::1#53(::1)
;; WHEN: Mon May 25 15:46:26 UTC 2020
;; MSG SIZE  rcvd: 59

$ dig www.les-play.com

; <<>> DiG 9.17.1-1+ubuntu16.04.1+isc+3-Ubuntu <<>> www.les-play.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 29388
;; flags: qr rd ra; QUERY: 1, ANSWER: 8, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.les-play.com.              IN      A

;; ANSWER SECTION:
www.les-play.com.       293     IN      CNAME   www.les-play.com.
www.les-play.com.       293     IN      CNAME   www.les-play.com.
www.les-play.com.       293     IN      CNAME   www.les-play.com.
www.les-play.com.       293     IN      CNAME   www.les-play.com.
www.les-play.com.       293     IN      CNAME   www.les-play.com.
www.les-play.com.       293     IN      CNAME   www.les-play.com.
www.les-play.com.       293     IN      CNAME   www.les-play.com.
www.les-play.com.       293     IN      CNAME   www.les-play.com.

;; Query time: 1983 msec
;; SERVER: ::1#53(::1)
;; WHEN: Mon May 25 15:46:35 UTC 2020
;; MSG SIZE  rcvd: 157
```

(It takes about 6 ms when trace-regex is off.)

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

A more fundamental fix to CNAME loops. Fixes #9153. We scan the names in the answer section for CNAMEs we've already seen.

Also see  #9194,  the throw in the CNAME cache lookup is needed to actually abort things.

Needs tests.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [X] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/syncres.cc [hunk 1]
```diff
@@ -3379,6 +3390,17 @@ bool SyncRes::processAnswer(unsigned int depth, LWResult& lwr, const DNSName& qn
       return true;
     }
 
+    // Check to see if we already have seen the new target as a previous target
```
Comment (pieterlexis) [inline pdns/syncres.cc:3393]: Splitting this out into a function that returns a bool might make sense.

File: pdns/syncres.cc [hunk 2]
```diff
@@ -1331,9 +1331,20 @@ bool SyncRes::doCNAMECacheCheck(const DNSName &qname, const QType &qtype, vector
...
+        throw ImmediateServFailException(msg);
+      }
+
+      // Check to see if we already have seen the new target as a previous target
+      for (const auto &rec: ret) {
+        if (rec.d_type == QType::CNAME && rec.d_place == DNSResourceRecord::ANSWER) {
+          if (newTarget == rec.d_name) {
+            string msg = "got a CNAME referral (from cache) that causes a loop";
+            LOG(prefix<<qname<<": status="<<msg<<endl);
+            throw ImmediateServFailException(msg);
```
Comment (pieterlexis) [inline pdns/syncres.cc:1345]: This throw means the behaviour of a cached vs uncached CNAME differs. The un-cached variant _will_ be a SERVFAIL with the CNAMEs in the response, where as the cached one won't be.

Comment (omoerbeek) [reply pdns/syncres.cc]: I checked what other resolvers do and hey return an empty answer RRSet. So I changed the code to do that as well.

**Comment by omoerbeek** [comment]
@pieterlexis I think your concerns are addressed. OK?

**Pull request merged**